Adding STAR

fhartley101 opened the pull request:
Adds STAR
1.  STAR index called in check and prepare input
2. STAR quant
3. FeatureCounts, which quantifies star alignments
4. genecounts which compiles the featurecounts output into a single csv

Modifies where the index files are stored (i.e. removes params.species from filepaths) to tidy output. Now indexes are stored in
results/salmon/index
results/star/index

Updated readme with new updates

fhartley101 committed:
adding star

alebarberis commented on config/modules.config:
```diff
@@ -93,6 +93,50 @@ process {
         ]
     }
 
+    thName: 'STAR_INDEX' {
```
Change to "withName"

alebarberis commented on modules/genecounts/main.nf:
```diff
@@ -0,0 +1,44 @@
+    //Set to true to forward process stdout to the current top stdout
+    debug false
+    //Associate process execution with custom label
+    tag "TXIMPORT_SALMON"
```
Check the tag.

alebarberis commented on modules/genecounts/main.nf:
```diff
@@ -0,0 +1,44 @@
+    //Annotate process with identifier
+    label "process_medium"
+    //Process dependencies
+    //conda "bioconda::bioconductor-tximeta=1.12.0"
```
You can have a conda container for pandas as well, like conda "conda-forge::pandas=2.2.2". You need to check the version though.

fhartley101 replied:
There is the 1.5.2 version on the anaconda website but it doesn't look like it's part of conda forge. Would the setup for this be "cctbx202211::pandas" ? I never managed to get conda working on my laptop so i can't test this
https://anaconda.org/cctbx202211/pandas

alebarberis commented on nextflow.config:
```diff
@@ -53,11 +53,13 @@ params {
     genemap         = null  // Path to file containing a mapping of transcripts to genes (tx2gene)
     salmon_index    = null  // Path to directory
+    star_index      = ""    // Path to directory
     decoys          = null  // Path to file
```
Maybe better to keep default values uniform and set star_index = null. If you agree and change it, make sure the change is not affecting other parts of the code.

alebarberis commented on subworkflows/check_and_prepare_input.nf:
```diff
@@ -324,16 +326,13 @@ workflow CHECK_AND_PREPARE_INPUT {
                         } else {
-
-
+                                if(params.decoys instanceof String){
+                                        fn_decoys = file(params.decoys).name
+                                } else {
+                                        fn_decoys = "decoys.txt"
+                                }
```
Please, check indentation.

alebarberis commented on subworkflows/check_and_prepare_input.nf:
Since the download of genome, transcriptome, and gtf files is a step shared by different quantification methods, we might want to think whether would be good to put it just after the CHECK REFERENCE section. For example, we could use something like: 

```{code}
if( (modules_to_run.contains('salmon') & params.salmon_index == true) | 
   (modules_to_run.contains('star') & params.star_index == true) ) {
   
   // Check transcriptome
  if( download_transcriptome ){
      //Download
      log.info "User-provided transcriptome file not found: try to download it"
      // Check species
      if( !params.species ){
          exit 1, "ERROR: `species` not specified!"
      }
      // Get URL
      def url = getTranscriptomeUrl(params.species)
      // Download file
      transcriptomeFilePath = downloadFile(
          url,
          refdir,
          null,
          true
      )
      // Update channel
      ch_transcriptome = Channel
          .fromPath(transcriptomeFilePath, checkIfExists: true, type: 'file')
  }
  
  
  // Check genome
  if( download_genome ){
      //Download
      log.info "User-provided genome file not found: try to download it"
      // Check species
      if( !params.species ){
          exit 1, "ERROR: `species` not specified!"
      }
      // Get URL
      def url = getGenomeUrl(params.species)
      // Download file
      genomeFilePath = downloadFile(
          url,
          refdir,
          null,
          true
      )
      // Update channel
      ch_genome = Channel
          .fromPath(genomeFilePath, checkIfExists: true, type: 'file')
  }
  
  
  // Check GTF
  if( download_gtf ){
      log.info "User-provided GTF file not found: try to download it"
      // Check species
      if( !params.species ){
          exit 1, "ERROR: `species` not specified!"
      }
      // Get URL
      def url = getGtfUrl(params.species)
      // Download file
      gtfFilePath = downloadFile(
          url,
          refdir,
          null,
          true
      )
      // Update channel
      ch_gtf = Channel
          .fromPath(gtfFilePath, checkIfExists: true, type: 'file')
  }
  
}
```

fhartley101 replied:
I think this might complicate things as the files aren't needed necessarily by both processes. For example, the transcriptome is never required by STAR, but always reuiqred by salmon. The genome is only required if you're building the index in both cases, whereas the GTF file is always required by STAR as it's needed for the quant step as well as the index generation but salmon only requires the GTF if the genemap option is selected.

alebarberis replied:
Ok, let's keep it as you modified then.

alebarberis commented on subworkflows/check_and_prepare_input.nf:
```diff
@@ -361,17 +360,35 @@ workflow CHECK_AND_PREPARE_INPUT {
+                            salmon_index
+                        )
+                        // Update
+                        salmon_index = GUNZIP.out.decompressed //possible issue here - should this be ch_salmon_index, or do we want it to go through the next steps?
```
No, it is correct as it is. We decompress the salmon_index, assign the output to salmon_index and then we check if salmon_index is a directory, and if the directory contains files. If it is so, we assign salmon_index to the ch_salmon_index, otherwise we raise an error.

fhartley101 replied:
Ah yes i see now!

fhartley101 committed:
Update modules.config

fixed "withName"

fhartley101 committed:
Update main.nf

fixed tag

fhartley101 committed:
Update check_and_prepare_input.nf

fixed indentation

fhartley101 committed:
Update main.nf

added conda pandas - note the version is different and i haven't tested this

fhartley101 committed:
Update nextflow.config

Changed "" to null when initialising star_index
I tested and it works with null so I don't know why I didn't do that to start with - makes sense to change it to null and keep consistent
